### PR TITLE
Fix: Get valid API version from error code for `resource group template deployment` resource when original API version is mismatched

### PR DESCRIPTION
**Pull request by ms-zhenhua**
https://github.com/hashicorp/terraform-provider-azurerm/issues/14810
image

### CODE REVIEW

File: internal/services/resource/template_deployment_common.go [hunk 1]
```diff
@@ -139,6 +140,25 @@ func filterOutTemplateDeploymentParameters(input interface{}) interface{} {
 	return output
 }
 
+func deleteResourceById(ctx context.Context, resourcesClient *resources.Client, resourceID string, resourceProviderApiVersion string) (result resources.DeleteByIDFuture, err error) {
```
Comment (katbyte) [inline internal/services/resource/template_deployment_common.go:143]: why are we creating a function when this is only used one place? could we please inline it?

**Comment by ms-zhenhua** [comment]
Hi @katbyte , thank you for reviewing. I have updated the code. Kindly help review it again.

**Comment by ms-zhenhua** [comment]
Hi @katbyte , I decided to add test case for this issue at first. However, there is a problem: the **azurerm_resource_group_template_deployment** for **softwareUpdateConfigurations** can only be configured to the VMs which have already reported data to the **azurerm_automation_account**. However, the VMs won't report data immediately after the **azurerm_virtual_machine_extension** is configured. It will cost a few minutes for VMs to report the data. Then the **azurerm_resource_group_template_deployment** can be successfully configured. 
image

Since Azure RM Provider does not have resource like **time_sleep** to set a time delay, I can only test it locally with **Time Provider** and it passed. The following configuration is what I used to test this issue. The test steps are:
1. Run `terraform apply`
2. Run `terraform destroy`


```
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "test" {
  name     = "demo-rg"
  location = "westus2"
}

# Create a Automation Account
resource "azurerm_automation_account" "test" {
  name                = "demo-aa"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  sku_name = "Basic"
}

# Create a Log Analytics Workspace
resource "azurerm_log_analytics_workspace" "test" {
  name                = "demo-law"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  sku                 = "PerGB2018"
  retention_in_days   = 30
}

# Link Log Analytics Workspace to Automation Account
resource "azurerm_log_analytics_linked_service" "test" {
  resource_group_name = azurerm_resource_group.test.name
  workspace_name      = azurerm_log_analytics_workspace.test.name
  resource_id         = azurerm_automation_account.test.id
}

# Enable Update Management solution
resource "azurerm_log_analytics_solution" "update_solution" {
  depends_on = [
    azurerm_log_analytics_linked_service.test
  ]

  solution_name         = "Updates"
  location              = azurerm_resource_group.test.location
  resource_group_name   = azurerm_resource_group.test.name
  workspace_resource_id = azurerm_log_analytics_workspace.test.id
  workspace_name        = azurerm_log_analytics_workspace.test.name

  plan {
    publisher = "Microsoft"
    product   = "OMSGallery/Updates"
  }
}

resource "azurerm_virtual_network" "test" {
  name                = "demo-vnet"
  address_space       = ["10.0.0.0/16"]
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
}

resource "azurerm_subnet" "test" {
  name                 = "demo-subnet"
  resource_group_name  = azurerm_resource_group.test.name
  virtual_network_name = azurerm_virtual_network.test.name
  address_prefix       = "10.0.2.0/24"
}

resource "azurerm_network_interface" "test" {
  name                = "demo-ni"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name

  ip_configuration {
    name                          = "testconfiguration1"
    subnet_id                     = azurerm_subnet.test.id
    private_ip_address_allocation = "Dynamic"
  }
}

resource "azurerm_virtual_machine" "test" {
  name                  = "demo-vm"
  location              = azurerm_resource_group.test.location
  resource_group_name   = azurerm_resource_group.test.name
  network_interface_ids = [azurerm_network_interface.test.id]
  vm_size               = "Standard_F4"

  storage_image_reference {
    publisher = "MicrosoftWindowsServer"
    offer     = "WindowsServer"
    sku       = "2016-Datacenter"
    version   = "latest"
  }

  storage_os_disk {
    name              = "myosdisk1"
    caching           = "ReadWrite"
    create_option     = "FromImage"
    managed_disk_type = "Standard_LRS"
  }

  os_profile {
    computer_name  = "zabc"
    admin_username = "zabc"
    admin_password = "Password1234!"
  }

  os_profile_windows_config {
    timezone           = "Pacific Standard Time"
    provision_vm_agent = true
  }
}

resource "azurerm_virtual_machine_extension" "test" {
  name                 = "demo-vme"
  virtual_machine_id   = azurerm_virtual_machine.test.id
  publisher            = "Microsoft.EnterpriseCloud.Monitoring"
  type                 = "MicrosoftMonitoringAgent"
  type_handler_version = "1.0"
  auto_upgrade_minor_version = "true"
  settings = <<SETTINGS
    {
      "workspaceId": "${azurerm_log_analytics_workspace.test.workspace_id}"
    }
SETTINGS
   protected_settings = <<PROTECTED_SETTINGS
   {
      "workspaceKey": "${azurerm_log_analytics_workspace.test.primary_shared_key}"
   }
PROTECTED_SETTINGS
}

resource "time_sleep" "wait_5_minutes" {
  depends_on = [
    azurerm_virtual_machine.test,
    azurerm_virtual_machine_extension.test
  ]

  create_duration = "300s"
}

resource "azurerm_resource_group_template_deployment" "test" {
  name                = "demo-deployment"
  resource_group_name = azurerm_resource_group.test.name
  deployment_mode     = "Incremental"
  depends_on = [
    time_sleep.wait_5_minutes
  ]

  template_content = <<TEMPLATE
  {
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "resources": [
        {
            "type": "Microsoft.Automation/automationAccounts/softwareUpdateConfigurations",
            "apiVersion": "2017-05-15-preview",
            "location": "${azurerm_resource_group.test.location}",
            "name": "${azurerm_automation_account.test.name}/demo-uc",
            "properties": {
                "updateConfiguration": {
                    "operatingSystem": "Windows",
                    "duration": "PT2H0M",
                    "windows": {
                      "includedUpdateClassifications": "Critical",
                      "rebootSetting": "IfRequired"
                    },
                    "azureVirtualMachines": [
                        "${azurerm_virtual_machine.test.id}"
                    ]
                },
                "scheduleInfo": {
                  "frequency": "Hour",
                  "startTime": "2022-10-19T12:22:57+00:00",
                  "timeZone": "America/Los_Angeles",
                  "interval": 1,
                  "expiryTime": "2022-11-09T11:22:57+00:00",
                  "advancedSchedule": {
                    "weekDays": [
                      "Monday",
                      "Thursday"
                    ]
                  }
                }
            }
        }
    ]
  }
  TEMPLATE
}
```

**Comment by nvukas59** [comment]
@katbyte could you please review this PR ?

**Comment by markpadam** [comment]
@katbyte any update on this one?

**Review by katbyte — APPROVED** [review]
LGTM 🦀

**Pull request merged**